Updating code for new ka-lint rules

JJC1138 opened the pull request:
Hello!

I'd like to apologise straight away for how big this patch is. With hindsight I should have pushed in smaller chunks every week or so, but it always looked like I was nearly done so I put it off. If you want to cherry pick commits rather than reviewing the whole lot in one go then please feel free, and I'll continue to merge or rebase any master changes to keep it current. Basically, if there's anything I can do to make reviewing this monster less of a burden, please let me know! And sorry again!

As you can see, I've done one commit per lint rule so that it's easier to review the changes. Many of the commits are purely mechanical formatting changes, of course.

There's one actual, real-life bug fix, which is in [Fix CSProgram.simpleValidate()](https://github.com/Khan/perseus/commit/6e83de19a167314041804bb02c59c5ea61960e1d).

[Fix some accidental globals that should be locals](https://github.com/Khan/perseus/commit/ac2b28b60f613f4a6a66699bb393f96670b19307) fixes some accidental-global errors, but as far as I could tell none of those actually caused problems (except perhaps for memory leaks).

Probably the largest part of the work was documenting prop types. I did the vast majority of them, but I did defer a few where the type was complex and it would have been very time consuming to figure out its complete specification. For those deferred ones I [inlined the `eslint-disable` comment](https://github.com/Khan/perseus/commit/b3f4d5b90861168c39636263f0b105fd905ce44f) around the prop in question so that the rest of file is lint clean. I adopted a convention of [using `React.PropTypes.shape({})`](https://github.com/Khan/perseus/commit/0e868fd73af4151f0e2b649c821e73fac471b430) to mean that the type can't be specified more precisely because it varies depending on how the component is used, and using `React.PropTypes.object` with the inlined `eslint-disable` comment to mean that the type could be more precisely defined but it just hasn't been done yet.

I think that's just about all there is to say. There are comments in some of the commit descriptions to give more detail where what I was doing wasn't obvious. I tested by running the test suite and by playing with every type of widget (including the hidden ones) and I wasn't able to find any breakages.

Thank you!

JJC1138 commented:
(This was previously discussed a bit in issue #65)

xymostech commented:
This looks awesome! I'm looking through it now :) (If I comment on individual commits, do those comments get put here? I guess we'll find out...)

xymostech commented on src/widgets/categorizer.jsx:
```diff
@@ -1,7 +1,9 @@
+/* eslint-disable no-var, react/jsx-closing-bracket-location, react/jsx-indent-props, react/prop-types, react/sort-comp */
 /* To fix, remove an entry above, run ka-lint, and fix errors. */
 
+/* global i18n:false */
```
Why not just `/* global i18n */`?

JJC1138 replied:
The `:false` tells ESLint that the global shouldn't be modified.

xymostech commented:
Could you split out the `CSProgram.simpleValidate()` fix into a separate commit?

xymostech commented on src/widgets/unit.jsx:
```diff
@@ -21,7 +21,6 @@ var MathOutput   = require("../components/math-output.jsx");
 var { SignificantFigures, displaySigFigs } = require("../sigfigs.jsx");
 
 var ALL = "all";
-var SOME = "some";
```
I feel like this is still useful but I don't know what it's for.

JJC1138 replied:
That one was used in the editor but not the widget itself if I remember correctly.

xymostech commented on src/widgets/plotter.jsx:
```diff
@@ -42,12 +42,12 @@ const widgetPropTypes = {
 
 var Plotter = React.createClass({
-    propTypes: {
+    propTypes: _.extend({
```
:( thanks for the fix, still sad that lint chokes

xymostech commented on src/widgets/lights-puzzle.jsx:
```diff
@@ -45,6 +43,7 @@ var RESET_BUTTON_STYLE = {
 
 var MAIN_TILE_SIZE = 50;
 
+/* eslint-disable no-unused-vars */
```
what's going on here?

JJC1138 replied:
Those were some unused functions that still looked useful so I didn't think I should delete them in case they were used for debugging or something?

xymostech commented on src/widgets/__tests__/expression-test.jsx:
```diff
@@ -150,6 +150,7 @@ var renderQuestionArea = function(item, apiOptions, enabledFeatures) {
     return renderer;
 };
 
+/* eslint-disable no-unused-vars */
```
what's going on here?

JJC1138 replied:
That's another unused function that I figured was probably used for debugging.

xymostech commented on src/widgets/transformer-editor.jsx:
```diff
@@ -659,15 +659,15 @@ var ShapeTypes = {
-            var line1_2 = [points1[1], points1[2]];
-            var line2_0 = [points2[1], points2[0]];
-            var line2_2 = [points2[1], points2[2]];
+            var line1a = [points1[1], points1[0]];
```
Hmm, kinda silly that it yells about this. It's hard to capitalize a zero! I'd vote for just disabling the lint rule, since the original syntax made more sense.

JJC1138 replied:
Yeah, I agree that's silly. I tried it with lint exception comments but I thought it looked marginally better this way.

xymostech commented on src/editor.jsx:
```diff
@@ -311,8 +315,22 @@ var imageUrlsFromContent = function(content) {
+        className: React.PropTypes.string,
+        content: React.PropTypes.string,
+        disabled: React.PropTypes.bool,
+        // We don't use EnabledFeatures.propTypes here because it requires the
```
:((( we should fix this, thanks for the note

ariabuckles replied:
This is/was a bit hard to fix because each key individually on enabledFeatures is .required, but overall as a whole enabledfeatures should be optional. Maybe it's possible to denote this correctly now! (It may also have been before and I assumed otherwise? Or maybe the defaulting logic for enabledFeatures is complicated [seems reasonably likely but i can't remember?])

JJC1138 replied:
I was hoping that there was a way of saying like `.isNotRequired` that would override something that was `.isRequired` but couldn't find any such thing.

I suppose I could just take the `.isRequired` off of EnabledFeatures.propTypes, and then add back the `.isRequired` inline everywhere that it's currently used. Would that be good?

ariabuckles replied:
if that works I think that's the right solution! (I think/hope it will work!)

JJC1138 replied:
Okay I'll give it a go. Thanks!

xymostech commented on src/components/sortable.jsx:
```diff
@@ -255,7 +255,7 @@ var Sortable = React.createClass({
         padding: React.PropTypes.bool,
         disabled: React.PropTypes.bool,
-        constraints: React.PropTypes.object,
+        constraints: React.PropTypes.shape({}),
```
I think I would prefer these were replaced with `React.PropTypes.any` if we don't know what the contents are. This seems to imply that they should take an object with no properties, or that we don't use any of the properties, whereas `any` says we don't know.

JJC1138 replied:
The reason I did that was that `any` can be literally anything like a number or whatever, whereas `shape({})` does at least says that it's an `object` so it's a _tiny_ bit more specific, but I'm happy to change those to `any`s if you'd prefer?

ariabuckles replied:
Did react get rid of React.PropTypes.object?

ariabuckles replied:
It's still on the react docs and if so, `React.PropTypes.object` specifies it's an object with unspecified keys, which is more specific than `any`, and should allow at least `undefined` unless we use `.isRequired` (i'm not sure if react still counts `null` as valid for non `.isRequired` propTypes).

JJC1138 replied:
`.object` is forbidden by the lint rule, so this was a bit of a hacky way around that tbh. I think `any` is probably better.

ariabuckles replied:
objects are totally valid and i object to that rule :/

JJC1138 commented:
I split out the `CSProgram.simpleValidate()` fix into PR #69. I'll rebase this branch when that's merged to master.

xymostech commented on src/editor-page.jsx:
```diff
@@ -17,7 +17,9 @@ var ViewportResizer = require("./components/viewport-resizer.jsx");
 var EditorPage = React.createClass({
     propTypes: {
+        /* eslint-disable react/forbid-prop-types */
         answerArea: React.PropTypes.object,
```
Similar to what I said in the previous commit, I think I'd rather these be `React.PropTypes.any` to show that we don't really know what's in them.

JJC1138 replied:
The ones I marked as shape({}) are unknowable, where as these are knowable but just unknown right now (</Rumsfeld>) and I think it's valuable to document the distinction, but it's probably better if I just do that with TODO comments.

xymostech commented on src/editor.jsx:
```diff
@@ -565,9 +565,6 @@ var Editor = React.createClass({
         localStorage.perseusLastCopiedText = selectedText;
         localStorage.perseusLastCopiedWidgets = JSON.stringify(widgetData);
-
-        console.log(
```
@alopatin I don't really know how perseus copying works, would this be useful in prod?

ariabuckles replied:
iirc widget copying was implemented as an intern project and i'm 90% sure this is just for debugging. it's maybe nice if a dev wants to inspect what happened on the perseus clipboard in the terminal, but I'd bet we don't actually do that.

I think it's okay to get rid of this.

xymostech commented on src/__tests__/perseus-markdown-test.jsx:
```diff
@@ -63,8 +63,10 @@ var htmlFromMarkdown = function(source) {
 var assertParsesToReact = function(source, html) {
     var actualHtml = htmlFromMarkdown(source);
     if (actualHtml !== html) {
+        /* eslint-disable no-console */
```
Somewhere we have a separate configuration for things that are run in a `node` environment, where `console.*` are fine. We should really be using that for things in `__tests__` instead of whitelisting these. But that's for another day :)

JJC1138 replied:
Oh yeah, I didn't think of that. Yeah, there's a separate `eslintrc.node` in khan-linter, but I don't know what the mechanism would be for specifying that for individual files.

JJC1138 replied:
There's a [TODO about that in khan-linter](https://github.com/Khan/khan-linter/blob/master/linters.py#L478) so it looks like that's not implemented there yet.

xymostech commented on src/widgets/cs-program-editor.jsx:
```diff
@@ -181,9 +181,12 @@ var CSProgramEditor = React.createClass({
             })
             .fail((jqxhr, textStatus, error) => {
+                /* eslint-disable no-console */
                 console.error("Error retrieving scratchpad info for " +
```
Not really sure this is a useful error, you could find the same thing in the network panel. I'd vote for nixing this.

xymostech commented on src/widgets/transformer.jsx:
```diff
@@ -460,7 +460,9 @@ var Transformations = {
             },
             componentDidUpdate: function(prevProps) {
                 if (!deepEq(this.props, prevProps)) {
+                    /* eslint-disable react/no-did-update-set-state */
```
:P I'm still not super happy with this lint rule, but this works.

xymostech commented:
Okay, I looked (skimmed?) through it all, let me know if you have any questions. Thank you so much for helping us out with this!!

JJC1138 commented:
You're very welcome! Thank you for checking it out so quickly and for the comments.

I'll start a provisional TODO list:
- [ ] add lint exception instead of changing `line1_0` to `line1a` etc.
- [ ] use `any` prop type instead of `shape({})` and `object` with comments distinguishing knowable versus unknowables
- [ ] rebase on PR #69 when merged
- [ ] cut the `console.error` in `cs-program-editor.jsx`

It's late here and I'm afraid I'm busy for the next few days, but I'll get to it on Monday. Thank you again! :)

JJC1138 commented:
Hello again. I've repushed with the changes we discussed.

I did experiment with making `ApiOptions.propTypes` and `EnabledFeatures.propTypes` non-required as me and @ariabuckles talked about, but it did turn out to be more complicated, because some option/feature props are required in some places and not others, so attempting to do it in a blanket way like that won't work.

Thanks!

ariabuckles commented:
Aha! So there *was* a reason for my madness (when I wrote that originally). Welp.

JJC1138 commented:
I rebased off master. No changes were required, but I built and added the new `build/` files now to the last commit since they're now checked-in. I didn't add the built files for all the other commits but can do if that's useful. Thanks!

xymostech commented:
@JJC1138 Sorry I haven't been terribly responsive about this pull request! I think that this all looks good for merging in. I don't know if anyone else wants to take a look through it, but that's okay.

Also @alopatin or @ariabuckles do you have opinions about whether we should squash these commits or not? I guess I'd prefer to squash just so there's less commits to page through/bisect over, but I don't have a terribly strong opinion.

JJC1138 commented:
@xymostech Oh no problem, I appreciate it was a bit of a large patch to deal with!

JJC1138 commented:
I just rebased again (the Travis build is failing because of #76, but it's building fine locally).

I'm moving on to another project at the moment so I might not have time to do this many more times, so if you can commit it soon that might be best to prevent it getting too crusty. If not, no hard feelings; I appreciate it's a big diff and no one asked me to do it :)

jeremywiebe commented:
Closing stale PR.

jeremywiebe commented:
Closing stale PR.
